feat: 회원의 비밀번호 제약 구현

minseok-oh opened the pull request:
## 📄 Summary

- 회원의 비밀번호 제약(글자수, 숫자, 영소문자)를 추가했습니다
- 기존의 테스트에서 비밀번호로 사용한 "password"를 "password00"으로 변경했습니다
- ValidatePasswordTest를 만들어 #104 와 연결시켰습니다

## 🙋🏻 More

아


close #104

minseok-oh committed:
feat: 회원가입 시 비밀번호 제약조건 코드 구현

- 사용자 도메인에 검증 메서드를 구현했습니다

minseok-oh committed:
refactor: 기존에 사용한 비밀번호를 password -> password00으로 변경

- 제약조건에 예외가 발생하지 않도록 조정했습니다

minseok-oh committed:
test: 비밀번호 제약조건인 글자수, 숫자, 소문자 예외를 담은 테스트 코드 작성

yudonggeun committed:
Merge branch 'dev' into feat/104-issue

chhs2131 commented:
아?! ㅋㅋㅋㅋㅋㅋㅋㅋㅌ

![image](https://github.com/user-attachments/assets/d596b384-5805-42ca-9d6f-16d5824d2be6)

yudonggeun commented on src/main/java/com/wootecam/luckyvickyauction/core/member/domain/Member.java:
```diff
@@ -19,6 +21,29 @@ public Member(Long id, String signInId, String password, Role role, Point point)
+            throw new BadRequestException("비밀번호는 8자 이상 20자 이하로 입력해주세요.", ErrorCode.M006);
+        }
+
+        if (!password.matches(".*[0-9].*")) {
+            throw new BadRequestException("비밀번호는 숫자가 반드시 포함되어야 합니다.", ErrorCode.M007);
+        }
+
+        if (!password.matches(".*[a-z].*")) {
```
`matches` 메서드는 내부적으로 `Pattern`을 사용하는 것으로 알고있는데요. `Pattern` 객체를 생성하는 비용이 비싸서 좋지 않다고 생각합니다. Pattern을 미리 캐싱하는 것은 어떨까요?

#### 추천 방안 (택 1)
* Member에 static 변수로 캐싱하기
* Validator 클래스를 새로 만들기

HiiWee replied:
공감합니다~~ 저는 static 변수로 같이 두는것에 한 표 드리겠읍니다! ㅎㅎ

minseok-oh replied:
반영 완료했습니다! 감사합니다!

yudonggeun commented on src/test/java/com/wootecam/luckyvickyauction/core/member/domain/MemberTest.java:
```diff
@@ -16,26 +16,30 @@ class MemberTest {
+
     @Test
     void usePoint_포인트를_사용할_수_있다() {
         // given
         Member buyer = Member.builder()
                 .signInId("testId")
-                .password("password")
+                .password("password00")
```
테스트 변경점이 엄청 많네요 ㄷㄷ
너무 고생했어요!!!!!!

HiiWee replied:
내 똥인데 이거,, 죄송함다 ㅠ

chhs2131 replied:
으으으으 진짜 고생했네용 ㅠㅠ

chhs2131 commented on src/main/java/com/wootecam/luckyvickyauction/core/member/domain/Member.java:
```diff
@@ -19,6 +21,29 @@ public Member(Long id, String signInId, String password, Role role, Point point)
         this.password = password;
         this.role = role;
         this.point = point;
+
+        validatePassword(password);
+    }
```
의견: 멤버변수 초기화 이후에 지연검증을 한다면, 메서드의 인자로 password를 넘겨주지 않고 this.password를 사용해도 될 것 같아요!

HiiWee replied:
validatePassword는 인자로 초기화하는 과정이전에 빠르게 예외로 쳐내는것도 방법이 될 것 같아요!

minseok-oh replied:
아이디 검증 밑에 두었습니다! 감사합니다!

chhs2131 commented on src/main/java/com/wootecam/luckyvickyauction/core/member/domain/Member.java:
```diff
@@ -19,6 +21,29 @@ public Member(Long id, String signInId, String password, Role role, Point point)
+        validatePassword(password);
+    }
+
+    /**
+     * 비밀번호 유효성 검사 <br>
+     * 1. 비밀번호는 최소 8자, 최대 20자리이다. <br>
+     * 2. 비밀번호는 숫자가 반드시 포함되어야한다. <br>
+     * 3. 비밀번호는 알파벳 소문자가 반드시 포함되어야한다. <br>
+     * @param password 비밀번호
+     */
+    private void validatePassword(String password) {
+        if (password == null || password.length() < 8 || password.length() > 20) {
+            throw new BadRequestException("비밀번호는 8자 이상 20자 이하로 입력해주세요.", ErrorCode.M006);
+        }
+
+        if (!password.matches(".*[0-9].*")) {
+            throw new BadRequestException("비밀번호는 숫자가 반드시 포함되어야 합니다.", ErrorCode.M007);
+        }
+
+        if (!password.matches(".*[a-z].*")) {
+            throw new BadRequestException("비밀번호는 알파벳 소문자가 반드시 포함되어야 합니다.", ErrorCode.M008);
+        }
     }
```
의견: 한글이나 특수문자에 대한 제한은 따로 두지 않아도 될까요?

minseok-oh replied:
제한 완료했슴다! 아이디 검증 충돌도 해결했슴다!!

HiiWee reviewed (commented):
귣귣~~ 고생하셨어요! 의견 몇개 남겼으니 확인 부탁드립니당🤓

chhs2131 commented on src/test/java/com/wootecam/luckyvickyauction/core/member/domain/ValidatePasswordTest.java:
```diff
@@ -0,0 +1,40 @@
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+/**
+ * @see <a href="https://github.com/woowa-techcamp-2024/Team7-ELEVEN/issues/104">#104</a>
+ */
+public class ValidatePasswordTest {
+
```
제안: abstract로 두고 MemberTest의 하위로 옮겨도 좋을 것 같아요!

minseok-oh committed:
Merge branch 'dev' into feat/104-issue

minseok-oh committed:
Merge branch 'dev' into feat/104-issue

minseok-oh committed:
fix: PR 코드리뷰 의견 반영

- pattern을 static으로 분리해 처리했습니다
- 검증 부분을 생성자의 상단으로 위치 변경했습니다
- 영문자, 숫자 외의 문자를 제약조건에 두었습니다

minseok-oh committed:
fix: PR 코드리뷰 의견 반영

- pattern을 static으로 분리해 처리했습니다
- 검증 부분을 생성자의 상단으로 위치 변경했습니다
- 영문자, 숫자 외의 문자를 제약조건에 두었습니다

HiiWee reviewed (approved):
고생하셨습니다 ㅠㅠ 마지막 PR 뭐지~
